packages/oauth/oauth-provider/src/lib/util/time.ts

```diff
@@ -0,0 +1,29 @@
+import { Awaitable } from './type.js'
+
+/**
+ * Utility function to protect against timing attacks.
+ */
+export async function constantTime<T>(
+  delay: number,
+  fn: () => Awaitable<T>,
+): Promise<T> {
+  if (!(delay > 0) || !Number.isFinite(delay)) {
+    throw new TypeError('Delay must be greater than 0')
+  }
+
+  const start = Date.now()
+  try {
+    return await fn()
+  } finally {
+    const delta = Date.now() - start
+    if (delta < delay) {
+      await new Promise((resolve) => setTimeout(resolve, delay - delta))
+    } else {
+      // The delay is too short, let's wait for the next multiple of `delay`
+      // to avoid leaking information about the execution time.
+      await new Promise((resolve) =>
+        setTimeout(resolve, delay * Math.ceil(delta / delay)),
+      )
```

Comment:
This will end-up waiting in total `delta + n * delay`, though I think the intent is for it to wait `n * delay`.  It's possible that this doesn't need to be a special case from `delta < delay`— we still want to "round up" to the next delay multiple.